src/bundle/Controller/UniversalDiscoveryController.php

```diff
@@ -161,58 +154,71 @@ public function accordionGridViewAction(
         Request $request,
         int $locationId
     ): JsonResponse {
-        $location = $this->loadLocation($locationId);
-
         $limit = $request->query->getInt('limit', 25);
         $sortClauseName = $request->query->getAlpha('sortClause', self::SORT_CLAUSE_DATE_PUBLISHED);
         $sortOrder = $request->query->getAlpha('sortOrder', Query::SORT_ASC);
+        $rootLocationId = $request->query->getInt('rootLocationId', self::ROOT_LOCATION_ID);
 
         $sortClause = $this->getSortClause($sortClauseName, $sortOrder);
 
-        $columns = $this->getColumns($location, $limit, $sortClause, true);
+        $columns = $this->getColumns($locationId, $limit, $sortClause, true, $rootLocationId);
 
         return new JsonResponse([
             'breadcrumb' => array_map(
                 function (Location $location) {
                     return $this->getRestFormat($location);
                 },
-                $this->getBreadcrumbLocations($location)
+                $this->getBreadcrumbLocations($locationId)
             ),
             'columns' => $columns,
         ]);
     }
 
     private function getColumns(
-        Location $location,
+        int $locationId,
         int $limit,
         Query\SortClause $sortClause,
-        bool $gridView = false
+        bool $gridView = false,
+        int $rootLocationId = self::ROOT_LOCATION_ID
     ): array {
-        $breadcrumbLocations = $this->getBreadcrumbLocations($location);
-        $breadcrumbLocationsLast = count($breadcrumbLocations) - 1;
-
-        $columnLocations = $breadcrumbLocationsLast < 2
-            ? $breadcrumbLocations
-            : [
-                $breadcrumbLocations[0], // First
-                $breadcrumbLocations[$breadcrumbLocationsLast - 1], // Before last
-                $breadcrumbLocations[$breadcrumbLocationsLast], // Last
+        $location = $this->locationService->loadLocation($locationId);
+
+        $locationPath = $this->getRelativeLocationPath($rootLocationId, $location->path);
+        $locationPathCount = count($locationPath);
+
+        if ($locationPathCount > self::COLUMNS_NUMBER) {
+            $locationPathLast = $locationPathCount - 1;
+
+            $columnLocations = [
+                $locationPath[0], // First
+                $locationPath[$locationPathLast - 2],
+                $locationPath[$locationPathLast - 1],
+                $locationPath[$locationPathLast], // Last
             ];
+        } else {
+            $columnLocations = $locationPath;
+        }
+
+        $columnLocations = array_map(function ($id) {
+            return (int)$id;
+        }, $columnLocations);
```

Comment:
This could be moved to foreach loop body in https://github.com/ezsystems/ezplatform-admin-ui/pull/1199/files#diff-4d222e59b38dddd932c9f1ba8a217019R208

```php
foreach ($columnLocations as $columnLocationId) {
   $columnLocationId = (int)$columnLocationId;

   # ...
}
```